ansible/playbook/tasks/update.yml

```diff
@@ -133,6 +133,9 @@
       command: supervisorctl start dashboard-upgrade
       changed_when: True
 
+    - name: Rename Grafana community panels 
+      command: /usr/share/percona-dashboards/fix-panels.py
```

Comment:
It will be run concurrently with dashboard upgrade script. Is it safe? Why we don't combine them?

Reply:
I have checked the FB prepared and no collisions are detected.

Also it's possible to fix the reason of leftovers by modifying a pmm-update playbook.   
https://github.com/Percona-Lab/percona-images/compare/PMM-5655-Collect-additional-logs
https://github.com/Percona-Lab/pmm-submodules/pull/776

Do you think that it's better to do in this way?

Reply:
> I have checked the FB and no collisions are detected.

That's great, but one can't be sure that race condition is not present by running a build/test cycle once. Even by successfully running it 1000 times, you can only assume that it will run successfully for users too, but not proof it – and then we will spend weeks trying to debug problems like the one Steve is having.

By reading code, it is clear for me that there is a potential for data race: `rename_panels` function in [import-dashboards.py](https://github.com/percona/pmm-server/blob/7c8695e4a3be2c904e7c2f8585bf416aef1c3f26/import-dashboards.py#L248-L259) is running concurrently with a function of the same name in [fix-panels.py](https://github.com/percona/grafana-dashboards/blob/2ef0f15e23709ccef2b296ad9e4a06c17125b452/fix-panels.py#L13-L29) and change the same directory/files. Even if currently they do not actually run at the same time due to timings, etc, the potential is still there. How can we eliminate it? Well, the easiest way is to run them sequentially, one after one, by placing them in the same file and using synchronous calls. So why not do so?

> Also it's possible to fix the reason of leftovers by modifying a pmm-update playbook.
> https://github.com/Percona-Lab/percona-images/compare/PMM-5655-Collect-additional-logs
I'm not sure what you mean there: you say pmm-update, but then link to the percona-images repo (where playbooks are frozen for PMM 2).

Reply:
> By reading code, it is clear for me that there is a potential for data race: `rename_panels` function in [import-dashboards.py](https://github.com/percona/pmm-server/blob/7c8695e4a3be2c904e7c2f8585bf416aef1c3f26/import-dashboards.py#L248-L259) is running concurrently with a function of the same name in [fix-panels.py](https://github.com/percona/grafana-dashboards/blob/2ef0f15e23709ccef2b296ad9e4a06c17125b452/fix-panels.py#L13-L29) and change the same directory/files. Even if currently they do not actually run at the same time due to timings, etc, the potential is still there. How can we eliminate it? Well, the easiest way is to run them sequentially, one after one, by placing them in the same file and using synchronous calls. So why not do so?
> 

So we would get calling the same removing procedure twice one by one. 

It looks like the play in pmm-update (Add community panels) is run after the start of service dashboard-upgrade. It's not obvious in jenkins logs but seems it's true due to the final result.

![Screenshot_20200330_165655](https://user-images.githubusercontent.com/25184101/78230332-da744680-74d9-11ea-9633-18c0a5f4951d.png)

As you may see we have correct names and old names at the same time.
```
natel-discrete-panel 
NatelEnergy-grafana-discrete-panel-f2c7e64
```

import-dashboards.py script completely drops the folder plugins  so original names couldn't left.

How about to run fix-panels.py script before starting dashboard-upgrade service?
![Screenshot_20200402_122449](https://user-images.githubusercontent.com/25184101/78232468-f9c0a300-74dc-11ea-89d9-29663b799130.png)


> > Also it's possible to fix the reason of leftovers by modifying a pmm-update playbook.
> > https://github.com/Percona-Lab/percona-images/compare/PMM-5655-Collect-additional-logs
> 
> I'm not sure what you mean there: you say pmm-update, but then link to the percona-images repo (where playbooks are frozen for PMM 2).

It's percona-images.

Reply:
> How about to run fix-panels.py script before starting dashboard-upgrade service?

Yes, that will fix the problem.

I still don't understand why we can't merge fix-panels.py into dashboard-upgrade script so we have a single script that does everything related to dashboards upgrade, though.

Reply:
/cc @askomorokhov

Reply:
Here are installation steps that relate to the issue:
- Started playbook pmm2-docker/main.yaml
- ....
- Play  "Supervisord start"
  - Started dashboard-upgrade service
    - Checked that it's a fresh installation and performed import-dashboards.py script
      - Added and renamed plugins folders
- Play "Grafana | Add community panels"
  - Added plugins folders (Now we have whole bunch of correct and old folders)
- Started playbook tasks/update.yml
- ...
- Play "Restart services"
  - Started dashboard-upgrade service
    - Checked that it's NOT a fresh installation and stopped import-dashboards.py script

So there are two ways of fixing the issue:
- Remove Play "Grafana | Add community panels" in the frozen playbook
- Add fix-panels.py script into update playbook on any place

Reply:
Thank you, I understand the problem now.

> * Play "Supervisord start"
> * * Started dashboard-upgrade service

And the we don't wait for it to finish? That's problematic by itself – now several processes modify Grafana configuration simultaneously.

I wonder if we can disable `dashboard-upgrade`'s autostart, and then start it manually when we know it is safe, and then wait for it to finish.

> Remove Play "Grafana | Add community panels" in the frozen playbook

It was frozen for a very simple reason – changing it will not affect existing installations. That's why all changes should be done in update. So that's not a solution.

> Add fix-panels.py script into update playbook before starting dashboard-upgrade service

That will work, except that `dashboard-upgrade` is still running concurrently.

Please try (or just estimate how much work would be required) to disable `dashboard-upgrade` autostart in pmm.ini, and call it manually. In theory, that should fix the problem, but I don't know how much changes will be required.

Reply:
And please update FB – some PRs are merged already.

Reply:
> And please update FB – some PRs are merged already.

Done.

Reply:
> > * Play "Supervisord start"
> > * * Started dashboard-upgrade service
> 
> And the we don't wait for it to finish? That's problematic by itself – now several processes modify Grafana configuration simultaneously.
> 

Actually we are waiting for a finish of the service.
We have got next play in the same playbook.
```
    - name: Grafana                   | Wait for dashboards
      wait_for:
        path: /srv/grafana/PERCONA_DASHBOARDS_VERSION
        state: present
```

This file is created right before finishing importing-dashboards script.

...
> > Add fix-panels.py script into update playbook before starting dashboard-upgrade service
> 
> That will work, except that `dashboard-upgrade` is still running concurrently.
> 
> Please try (or just estimate how much work would be required) to disable `dashboard-upgrade` autostart in pmm.ini, and call it manually. In theory, that should fix the problem, but I don't know how much changes will be required.

IMHO it would take couple hours for preparing changes and much more hours for tests of updating procedure from different PMM versions.

Reply:
Oh, that's good that we wait for script to finish.

Then the only thing left if Travis CI build failure.